### PR TITLE
Grid: Convert component to TypeScript

### PR DESCRIPTION
**Pull request by walbo**
## What?
Converts the `Grid` component to TypeScript.

## Why?
Part of the @wordpress/components's TypeScript migration (https://github.com/WordPress/gutenberg/issues/35744).

## How?
- Refactors the current `Grid` to TypeScript.
- Refactors the tests to TypeScript.
- Refactors the storybook to TypeScript and removed knobs.

## Testing Instructions
1. Run npm run dev and ensure no typing errors
2. Fire up the Storybook and confirm the `Grid` continues to function as expected

### CODE REVIEW

File: packages/components/src/grid/README.md
Comment (aaronrobertshaw) [inline packages/components/src/grid/README.md]: Looks like we're missing the `columnGap` prop in the readme.

File: packages/components/src/grid/types.ts
Comment (aaronrobertshaw) [inline packages/components/src/grid/types.ts:24]: Given the use of the `GridColumns` and `GridRows` types, the auto-generated docs in Storybook are less helpful. Could we add something to the `columns` and `rows` prop descriptions to illustrate that they are in effect `number | number[]`?

Nit: We should probably fix up the alphabetization of the props as well by correcting the order of `isInline` and `gap`.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for working on this one @walbo 👍 

Everything tests as advertised for me. I left a couple of minor suggestions but am happy to approve if we decide the prop descriptions are ok as they are.

✅ No typing errors
✅ Unit tests pass
✅ Code changes make sense
✅ Has changelog
✅ Component functions correctly in both Storybook and editors
✅ Storybook example has appropriate controls
❔ README matches type props and descriptions
❔ Auto-generated docs in Storybook are meaningful

File: packages/components/src/grid/types.ts
Comment (mirka) [reply packages/components/src/grid/types.ts:24]: Wow, TIL this [`useResponsiveValue()`](https://github.com/WordPress/gutenberg/blob/212163897db9b1e4b68e04393e9f450808d4a3fc/packages/components/src/grid/hook.js#L41) construct.

I imagine it would be difficult for anyone to understand why `number[]` is an acceptable value, when there is no explanation anywhere of the `useResponsiveValue` thing. Before even thinking about how to document this feature properly, we might have to first reconsider whether we want to continue providing this feature officially 🤔 Like, how useful are viewport breakpoints in a component library like this? Should we not even bother until [container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Container_Queries) land? (cc @ciampo)

In the meantime, could we maybe just pretend, for the casual docs reader, that it only takes a number? 😂 

```js
	/**
	 * The number of columns in the `Grid`.
	 *
	 * @default 2
	 */
	columns?: ResponsiveCSSValue< number >;
```

(☝️ And I think we can bypass the `GridColumns`/`GridRows` type aliases, as it just adds another layer of indirection that doesn't mean much.)

File: packages/components/src/grid/types.ts
Comment (mirka) [inline packages/components/src/grid/types.ts]: For consistency, could we import the `ResponsiveCSSValue` alias from the utility types instead of duplicating the definition in this file?

https://github.com/WordPress/gutenberg/blob/212163897db9b1e4b68e04393e9f450808d4a3fc/packages/components/src/ui/utils/types.ts#L1

**Review by mirka — COMMENTED** [review]
Looks good in general!

TypeScript is telling me that `gridColumnGap`/`gridRowGap` are deprecated 😮 Apparently they were renamed in CSS3? Let's just fix that up for spec compliance. 

```diff
	/**
-	 * Adjusts the `grid-column-gap`.
+	 * Adjusts the `column-gap`.
	 */
-	columnGap?: CSSProperties[ 'gridColumnGap' ];
+	columnGap?: CSSProperties[ 'columnGap' ];
```

**Comment by aaronrobertshaw** [comment]
> TypeScript is telling me that gridColumnGap/gridRowGap are deprecated 😮 Apparently they were renamed in CSS3? Let's just fix that up for spec compliance.

I have a question about this. The `columnGap` and `rowGap` props are actually used to [create CSS rules](https://github.com/WordPress/gutenberg/blob/8faf87fa0bf1549b9534baeb0f1034c1cf23c68c/packages/components/src/grid/hook.ts#L61-L62) for `gridColumnGap` and `gridRowGap` respectively. Is it better to leave the type consistent with its actual use and make that runtime change separately in a follow-up?

This is assuming we're trying to strictly follow the guidelines in not making runtime changes during TypeScript refactors. It's only a very minor change of course, hence this just being a question 🙂

**Comment by mirka** [comment]
> I have a question about this. The `columnGap` and `rowGap` props are actually used to [create CSS rules](https://github.com/WordPress/gutenberg/blob/8faf87fa0bf1549b9534baeb0f1034c1cf23c68c/packages/components/src/grid/hook.ts#L61-L62) for `gridColumnGap` and `gridRowGap` respectively. Is it better to leave the type consistent with its actual use and make that runtime change separately in a follow-up?

Ah you're right, I was just thinking about types but it makes sense to change the CSS code as well. Let's move it to a follow-up then 👍

File: packages/components/src/grid/README.md
Comment (walbo) [reply packages/components/src/grid/README.md]: Will add later today 👍

File: packages/components/src/grid/stories/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/grid/stories/index.tsx]:
```suggestion
		columns: {
			table: { type: { summary: 'number' } },
			control: { type: 'number' },
		},
```

Should we label the `columns` prop type as `number` within the Storybook docs to match the readme as per: https://github.com/WordPress/gutenberg/pull/41923#discussion_r906097855

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the refinements to this @walbo 👍 

This is looking pretty good.

✅ No typing errors
✅ Unit tests pass
✅ Component functions well in editor and storybook

I'm happy to give this a tentative approval. My only remaining suggestion is in relation to the displayed type in the component's storybook docs.

> In the meantime, could we maybe just pretend, for the casual docs reader, that it only takes a number? 

This PR currently does pretend in the README that `columns` and `rows` only take `number` values. In the auto-generated storybook docs however it still shows as `ResponsiveCSSValue< number >`. 

Would it avoid some confusion and further the efforts to obfuscate the responsive values type if we updated the storybook controls table to only list `number` as the type as well? 

Screen Shot 2022-06-28 at 3 50 37 pm

File: packages/components/src/grid/stories/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/grid/stories/index.tsx]:
```suggestion
		rows: {
			table: { type: { summary: 'number' } },
			control: { type: 'number' },
		},
```

If updating `columns` this will need a tweak as well.

File: packages/components/src/grid/types.ts
Comment (ciampo) [reply packages/components/src/grid/types.ts:24]: > Before even thinking about how to document this feature properly, we might have to first reconsider whether we want to continue providing this feature officially 🤔 Like, how useful are viewport breakpoints in a component library like this? Should we not even bother until container queries land?

Good points.

Personally, I would:

- see if this feature is used at all in the codebase
- discuss/agree if this needed by the component in the first place
- expand the conversation to the package level — ie., do we need such a mechanism in our package? If so, should we look into implementing a more defined pattern for responsive layouts at all?

Personally, this feature feels a bit overkill. And if we zoom out even further, I'd like to take a more broad look at layout components like `Flex`, `Grid` and `{H|V|Z}Stack`, to asses their APIs and many any potential adjustments.

**Review by ciampo — COMMENTED** [review]
Not much to add to the very comprehensive reviews from @aaronrobertshaw  and @mirka !

> Let's move it to a follow-up then 👍

Agreed, let's follow-up with those changes to the `gridRowGap` and `gridColumnGap` props

**Comment by walbo** [comment]
Thanks for the feedback and reviews 👍

**Pull request merged**
